### PR TITLE
[IOTDB-6209] Pipe: Solving the topological order of the progress index in the historical data collection phase

### PR DESCRIPTION
**Pull request by yschengzi**
Problem:
When the pipe performs historical data collection, it currently sends sequential data first and then disorganized data, which is obviously wrong because the progress index of some of the disorganized data may be smaller than that of sequential files, which causes the pipe to record the wrong progress index as the progress information, resulting in some of the disorganized files not being sent.
Solution:
After collecting all the historical data, solve for the topological order according to the progress index contained in the TsFile. Use the topological order of the file as the order in which the historical data is collected.

问题：
在 pipe 进行历史数据收集的时候，目前是先发送顺序数据，再发送乱序数据，这明显是错误的，因为部分乱序数据的 progress index 可能比顺序文件更小，导致 pipe 会记录错误的 progress index 作为进度信息，导致部分乱序文件没有被发送。
解决方案：
在收集完所有历史数据之后，按照 TsFile 包含的 progress index 求解拓扑序。使用文件的拓扑序作为历史数据收集的顺序。

### CODE REVIEW

File: iotdb-core/datanode/src/test/java/org/apache/iotdb/db/storageengine/dataregion/TsFileResourceProgressIndexTest.java [hunk 1]
```diff
@@ -218,4 +229,100 @@ public void testHybridProgressIndex() {
     Assert.assertFalse(
         hybridProgressIndex.isAfter(new RecoverProgressIndex(1, new SimpleProgressIndex(2, 21))));
   }
+
+  @Test
+  public void testProgressIndexTopologicalSort() {
+    Random random = new Random();
+    List<ProgressIndex> progressIndexList = new ArrayList<>();
+
+    int ioTProgressIndexNum = 10, peerIdRange = 3, searchIndexRange = 1000;
```
Comment (SteveYurongSu) [inline iotdb-core/datanode/src/test/java/org/apache/iotdb/db/storageengine/dataregion/TsFileResourceProgressIndexTest.java:238]: Make all the number larger. Random number is recommanded.

File: iotdb-core/node-commons/src/main/java/org/apache/iotdb/commons/consensus/index/impl/MinimumProgressIndex.java [hunk 2]
```diff
@@ -79,6 +81,11 @@ public ProgressIndexType getType() {
     return ProgressIndexType.MINIMUM_PROGRESS_INDEX;
   }
 
+  @Override
+  public TotalOrderSumTuple getTotalOrderSumTuple() {
+    return new TotalOrderSumTuple(0L);
```
Comment (SteveYurongSu) [inline iotdb-core/node-commons/src/main/java/org/apache/iotdb/commons/consensus/index/impl/MinimumProgressIndex.java:86]:
```suggestion
    return new TotalOrderSumTuple(0L);
```

private static final TotalOrderSumTuple TOTAL..TUPLE = new TotalOrderSumTuple(Long.MIN_VALUE);

File: iotdb-core/node-commons/src/main/java/org/apache/iotdb/commons/consensus/index/ProgressIndex.java [hunk 3]
```diff
@@ -83,6 +92,8 @@ public interface ProgressIndex {
   /** @return the type of this progress index */
   ProgressIndexType getType();
 
+  TotalOrderSumTuple getTotalOrderSumTuple();
```
Comment (SteveYurongSu) [inline iotdb-core/node-commons/src/main/java/org/apache/iotdb/commons/consensus/index/ProgressIndex.java:95]: Add some docs. Describe limitation.

File: iotdb-core/datanode/src/main/java/org/apache/iotdb/db/pipe/extractor/historical/PipeHistoricalDataRegionTsFileExtractor.java [hunk 4]
```diff
@@ -287,6 +291,11 @@ && isTsFileGeneratedAfterExtractionTimeLowerBound(resource))
               }
             });
 
+        // progress index topological sort
+        resourceList.sort(
+            (Comparator.comparing(o -> o.getMaxProgressIndex().getTotalOrderSumTuple())));
```
Comment (SteveYurongSu) [inline iotdb-core/datanode/src/main/java/org/apache/iotdb/db/pipe/extractor/historical/PipeHistoricalDataRegionTsFileExtractor.java:296]:
```suggestion
        ProgressIndex.topologicalSort(resourceList);
```

File: iotdb-core/datanode/src/main/java/org/apache/iotdb/db/pipe/extractor/historical/PipeHistoricalDataRegionTsFileExtractor.java [hunk 5]
```diff
@@ -287,6 +291,11 @@ && isTsFileGeneratedAfterExtractionTimeLowerBound(resource))
               }
             });
 
+        // progress index topological sort
```
Comment (SteveYurongSu) [inline iotdb-core/datanode/src/main/java/org/apache/iotdb/db/pipe/extractor/historical/PipeHistoricalDataRegionTsFileExtractor.java:294]: remove comment

File: iotdb-core/datanode/src/main/java/org/apache/iotdb/db/pipe/extractor/historical/PipeHistoricalDataRegionTsFileExtractor.java [hunk 6]
```diff
@@ -287,6 +291,11 @@ && isTsFileGeneratedAfterExtractionTimeLowerBound(resource))
...
+        resourceList.sort(
+            (Comparator.comparing(o -> o.getMaxProgressIndex().getTotalOrderSumTuple())));
+        pendingQueue = new ArrayDeque<>(resourceList);
+
         LOGGER.info(
```
Comment (SteveYurongSu) [inline iotdb-core/datanode/src/main/java/org/apache/iotdb/db/pipe/extractor/historical/PipeHistoricalDataRegionTsFileExtractor.java:299]: add some processing time info here.

File: iotdb-core/node-commons/src/main/java/org/apache/iotdb/commons/consensus/index/ProgressIndex.java [hunk 7]
```diff
@@ -121,4 +132,58 @@ static ProgressIndex blendProgressIndex(
...
+  class TotalOrderSumTuple implements Comparable<TotalOrderSumTuple> {
+    private final ImmutableList<Long> tuple;
+
+    public TotalOrderSumTuple(Long... args) {
+      tuple = ImmutableList.copyOf(args);
+    }
+
+    public TotalOrderSumTuple(List<Long> list) {
+      tuple = ImmutableList.copyOf(list);
+    }
```
Comment (SteveYurongSu) [inline iotdb-core/node-commons/src/main/java/org/apache/iotdb/commons/consensus/index/ProgressIndex.java:149]:
```suggestion
    private TotalOrderSumTuple(Long... args) {
      tuple = ImmutableList.copyOf(args);
    }

    private TotalOrderSumTuple(List<Long> list) {
      tuple = ImmutableList.copyOf(list);
    }
```

File: iotdb-core/node-commons/src/main/java/org/apache/iotdb/commons/consensus/index/impl/MinimumProgressIndex.java (hunk 2 above)
Comment (yschengzi) [reply iotdb-core/node-commons/src/main/java/org/apache/iotdb/commons/consensus/index/impl/MinimumProgressIndex.java:86]: negative value of MinimumProgressIndex is forbidden for the correctness of algorithm.

File: iotdb-core/datanode/src/test/java/org/apache/iotdb/db/storageengine/dataregion/TsFileResourceProgressIndexTest.java (hunk 1 above)
Comment (yschengzi) [reply iotdb-core/datanode/src/test/java/org/apache/iotdb/db/storageengine/dataregion/TsFileResourceProgressIndexTest.java:238]: done.

File: iotdb-core/datanode/src/main/java/org/apache/iotdb/db/pipe/extractor/historical/PipeHistoricalDataRegionTsFileExtractor.java (hunk 5 above)
Comment (yschengzi) [reply iotdb-core/datanode/src/main/java/org/apache/iotdb/db/pipe/extractor/historical/PipeHistoricalDataRegionTsFileExtractor.java:294]: done.

**Comment by SteveYurongSu** [comment]
pull and merge master please

File: iotdb-core/datanode/src/main/java/org/apache/iotdb/db/pipe/extractor/historical/PipeHistoricalDataRegionTsFileExtractor.java (hunk 4 above)
Comment (yschengzi) [reply iotdb-core/datanode/src/main/java/org/apache/iotdb/db/pipe/extractor/historical/PipeHistoricalDataRegionTsFileExtractor.java]: done.

File: iotdb-core/node-commons/src/main/java/org/apache/iotdb/commons/consensus/index/ProgressIndex.java (hunk 3 above)
Comment (yschengzi) [reply iotdb-core/node-commons/src/main/java/org/apache/iotdb/commons/consensus/index/ProgressIndex.java]: done.

File: iotdb-core/node-commons/src/main/java/org/apache/iotdb/commons/consensus/index/impl/MinimumProgressIndex.java [hunk 8]
```diff
@@ -79,6 +81,11 @@ public ProgressIndexType getType() {
     return ProgressIndexType.MINIMUM_PROGRESS_INDEX;
   }
 
+  @Override
+  public TotalOrderSumTuple getTotalOrderSumTuple() {
+    return new TotalOrderSumTuple();
+  }
+
```
Comment (SteveYurongSu) [inline iotdb-core/node-commons/src/main/java/org/apache/iotdb/commons/consensus/index/impl/MinimumProgressIndex.java:88]:
```suggestion
  @Override
  public TotalOrderSumTuple getTotalOrderSumTuple() {
    return TOTAL_ORDER_SUM_TUPLE;
  }
  
  public static final MinimumProgressIndex INSTANCE = new MinimumProgressIndex();
  
  private static final TotalOrderSumTuple TOTAL_ORDER_SUM_TUPLE = new TotalOrderSumTuple();
```

**Review by SteveYurongSu — APPROVED** [review]
Make Pipe Great Again!!!

**Pull request merged**